---
title: "IBD segments in haplotype data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,warning = FALSE, message = FALSE,fig.retina = 2)
library(tidyverse)
library(wesanderson)
library(ComplexHeatmap)
```

Any two samples in a population can be traced back to their most recent common ancestor by sharing identity-by-descent haplotypes. Although we don't have a chromosome-level reference genome, the version two of *A.digitifera* display significant improvement in N50 size, thus it is worth investigating the haplotype segments sharing pattern in population.

### Infer identity-by-descent segemnts
We used `refined-ibd` [v17Jan20](https://faculty.washington.edu/browning/refined-ibd.html) to detect IBD segments in phased genotype data. The map file was generated as in [selection analysis](06.selection_analysis.md).

```bash
java -jar refined-ibd.17Jan20.102.jar nthreads=10 \
gt=Adigi.v2.indv74_phased.vcf.gz map=map.txt chrom={scaffold} \
length=0.15 trim=0.015 window=4 out={scaffold}
```

Next, we use `merge-ibd-segments` to remove any breaks and short gaps in IBD segments.
```bash
zcat {scaffold}.ibd.gz | java -jar merge-ibd-segments.17Jan20.102.jar \
{scaffold}.vcf map.txt 0.01 2 > {scaffold}.merged.ibd
```
### Calculate the relatedness
The pairwise relatedness can be calculated using a python script [relatedness_v1.py](http://faculty.washington.edu/sguy/ibd_relatedness.html) in which the relatedness was calculated as the proportion of shared haplotype length divided by the total chromosome length*2.
```bash
cat *.merged.ibd | python2 relatedness_v1.py map.txt 0 > Adigi_ind74.ibd_relatedness.txt
```

```{r ibd-relatedness-plot, fig.width= 6, fig.height=5}
relatedness <- read_tsv("data/hpc/ibd/Adigi_ind74.ibd_relatedness.txt") %>% column_to_rownames("ID") %>% as.matrix()

shortid <- rownames(relatedness) %>% str_split("_") %>% 
  map(~.x[1:3]) %>% map(paste,collapse="_" ) %>% 
  unlist
names(shortid) <- rownames(relatedness)

ids<-rownames(relatedness) 
population<-case_when(substr(ids,1,2)=="AR" ~ "NO",
          substr(ids,1,2)=="AI" ~"IN",
          substr(ids,1,2)=="BR" ~"IN",
          substr(ids,1,2)=="RS" ~"SO")

ha=HeatmapAnnotation(bar = population,
                     show_legend = FALSE,show_annotation_name = FALSE,
    col = list(bar = c("IN" = "#CC0C00", "NO" = "#5C88DA", "SO" = "#84BD00")))

pal <- wes_palette("Zissou1", 500, type = "continuous")
Heatmap(log(relatedness), column_labels = shortid,
        clustering_distance_columns="euclidean",
        show_column_dend = FALSE,
        show_column_names = FALSE,
        row_labels = shortid[rownames(relatedness)],
        row_names_gp = gpar(fontsize = 5),
        name="relatedness",top_annotation = ha,
        show_heatmap_legend = FALSE,
        col = pal)

```

**Figure:** A heatmap shows the pairwise relatedness based on shared haplotypes. A darker color means a higher relatedness.


