---
title: "Population Structure"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
library(tidyverse)
library(ggsci)
library(stringi)
library(purrr)
library(patchwork)
library(PCAviz)
source(file = "scripts/color_scheme.R")
```

Starting with a fully filtered VCF, we performed linkage pruning and prepared data for PCA, ADMIXTURE analysis.

### LD pruning

Because plink will only take common chromosome id like in human, we first created a map file for *Acropora digitifera*.

```bash
bcftools view -H Adigi.DPg90gdp3gq30.vcf.gz | \
  cut -f 1| uniq | \
  awk '{print $0"\t"$0 }' > Adigi.DPg90gdp3gq30.chrom-map.txt
```

Next, we converted our vcf file to plink format which included sites with a minor allele count greater than or equal to two and excluded sites with a p-value for exact test for Hardy-Weinberg Equilibrium below 1e-4. Eventually, we pruned SNPs in highly linked regions.

```bash
plink --vcf Adigi.v2.filtered.vcf.gz --out Adigitifera \
  --allow-extra-chr --make-bed --set-missing-var-ids @:# \
  --double-id --geno 0.1 --mac 2 --hwe 0.0001

plink --bfile Adigitifera --indep-pairwise 50 10 0.1 --out Adigitifera \
    --allow-extra-chr --set-missing-var-ids @:#
    
plink --bfile Adigitifera --extract Adigitifera.prune.in \
    --recode12 --allow-extra-chr --out Adigitifera_ldpruned
```
We then start with our PCA analysis and Admixture analysis with **919,129** SNPs.

### Principal Component Analysis
**Is there genetic structure in Kimberley corals?**

`smartpca` has a design flaw that it won't work with chromosome id other than small numbers (1-22,X,Y). So we used an [in-house script](scripts/fix_mapfile_for_smartpca.py) to fix the map file and run smartpca with default parameters.

```{r smartpca,fig.retina=2,fig.show="hold", out.width="100%",fig.cap="Two plots depict the first vs second principal components (PC1 vs PC2) and PC2 vs PC3, which stratified by sample site location (red-like dots: Inshore, green-like dots from south offshore, blue dots: north offshore)."}
pca <- read.table("data/hpc/pca/Adigitifera_smartpca.evec",
                       comment = "#",sep="",
                       col.names =c("sampleid",paste0("PC",1:20),"location"))

sdev_eval <- sqrt(scan("data/hpc/pca/Adigitifera_smartpca.eval")[1:20])
sum_eval <- sum(scan("data/hpc/pca/Adigitifera_smartpca.eval"))
pve <- sdev_eval/sum_eval*100

p1 <- ggplot(pca %>% arrange(desc(location)), aes(PC1,PC2,color=location)) +
  geom_point(size=2,alpha=0.8,shape=16)  +
  scale_color_manual(values = my_color) + 
  labs(x=paste0("PC1(", signif(pve[1],3),"%)"), y=paste0("PC2(", signif(pve[2],3),"%)")) + theme_bw() +
  theme(text=element_text(size=12,  face = "bold"),
        legend.text = element_text(size=8,face = "bold"),
        legend.title = element_blank(),
        legend.position = c(0.18,0.9),
        legend.box = "horizontal",
        legend.direction = "horizontal",
        panel.grid = element_blank()) 

p2 <- ggplot(pca %>% arrange(desc(location)), aes(PC2,PC3,color=location)) +
  geom_point(size=2,alpha=0.8,shape=16)  +
  scale_color_manual(values = my_color) + 
  labs(x=paste0("PC2(", signif(pve[2],3),"%)"), y=paste0("PC3(", signif(pve[3],3),"%)")) + theme_bw() +
  theme(text=element_text(size=12,  face = "bold"),
        legend.position = "none",
        panel.grid = element_blank()) +
  geom_label(data = pca %>% 
               extract(sampleid, "label", "([^\\:]*)", remove = FALSE) %>% 
               filter(grepl(pattern = "^BR_5_121",sampleid)), 
             aes(label=label), nudge_y = 0.1, size=2) 


p1 + p2
```


Corals from different regions were well separated as inshore, offshore north, and offshore south, except one sample from Beagle reef was clustered with offshore south. In the right plot, inshore samples were tightly congregated compared to two offshore groups. Therefore, our samples from six locations form three clusters which correspond to three geographic regions we called inshore, offshore north, and offshore south.

This analysis revealed that one sample from Beagle Reef (inshore), `BR_5_121` had offshore ancestry despite being collected from inshore.  We excluded this sample from subsequent analyses of demography and selection.

### PC loadings
To check whether population structure evident in the PCA was driven by a small number of genomic regions (eg such as inversions) we inspected the PC loadings for the first two principle components.

```{r pc_loading,cache=TRUE,fig.retina=2,fig.show="hold", out.width="50%"}
snpsigs <- read_table("data/hpc/pca/Adigitifera_smartpca.snpeigs",col_names = c("ID","chr","pos",paste0("PC",(1:20)))) 

hgdp <- pcaviz(dat=pca, sdev = sdev_eval, var = sum_eval, rotation = snpsigs ) 

pcaviz_loadingsplot(hgdp,pc.dim = "PC1",min.rank = 0.8,geom.point.params = list(size=0.5, color="grey36")) + labs(x="SNPs",y=paste0("PC",1)) + ggtitle("PC loading plots of SNPs across the genome")

pcaviz_loadingsplot(hgdp,pc.dim = "PC2",min.rank = 0.8,geom.point.params = list(size=0.5, color="grey36")) + labs(x="SNPs",y=paste0("PC",2)) 

```
At a high level this seems to suggest that population structure reflects variation across the entire genome rather and is not dominated by a small number of strongly differentiated regions.

### Admixture analysis

**Is there any mixture in samples?**

Using the unlinked SNPs in plink format we created before, we run `ADMIXTURE` with default parameters and 10-fold cross-validation. The postulated number of ancestral population K was set from 1 to 6 and get results on cross-validation error across values of K.

>`ADMIXTURE` will need genotypes coded as 0,1,2 (instead of A,T,G,C).

```bash
for K in {1..6}
do
  admixture --cv=10 Adigitifera_ldpruned.ped $K | tee log.${K}.out
  awk -v K=$K '$1=="CV"{print K, $4}' log.${K}.out >> CV.txt
done
```
The cross-validation error suggests the lowest cross-validation error is with K=1 which is a common situation when population differentiation is subtle. We then check results for K=2,3.

```{r admixture, fig.width=5.2, fig.height=3.8,fig.retina=2,out.extra='style="float:left"'}
samples <- scan("data/hpc/admixture/sample_order.txt", character())
read_Q <- function(filename) {
  tb <- read_table(filename,col_names = FALSE) %>% 
  add_column(samples=samples) %>% 
  mutate(pop=str_split(samples,"_") %>% map_chr(first)) %>% 
  gather(cluster,proportion,-samples,-pop)
  
  tb$samples <- factor(tb$samples, levels = tb %>% 
                         arrange(pop) %>% pull(samples) %>% unique())
  tb$pop <- factor(tb$pop,levels = c("AI","BR","AR","RS1","RS2","RS3"))
  
  tb$cluster <- factor(tb$cluster, levels=rev(unique(tb$cluster)))
  
  tb
}
pop.labs <- c("AI","BR","AR","RS1","RS2","RS3")
names(pop.labs) <- c("AI","BR","AR","RS1","RS2","RS3")
p1 <- ggplot(read_Q("data/hpc/admixture/Adigitifera_ldpruned.2.Q"),aes(x=samples)) + geom_col(aes(group=samples,y=proportion,fill=cluster),show.legend = FALSE) + 
  theme_minimal() + 
  facet_grid(~pop,switch = "x",scales = "free",space = "free") +
  scale_color_brewer(palette = "Set2") +
  theme(axis.text.x = element_blank(),panel.grid = element_blank(),strip.text.x = element_blank())+
  labs(y="Ancestry (K=2)", x="") 

p2<-ggplot(read_Q("data/hpc/admixture/Adigitifera_ldpruned.3.Q"),aes(x=samples)) + geom_col(aes(group=samples,y=proportion,fill=cluster),show.legend = FALSE) + theme_minimal() +
  facet_grid(~pop,switch = "x",scales = "free",space = "free",
             labeller = labeller(pop=pop.labs)) +
  scale_fill_brewer(palette = "Set2",labels= c("X1"="South Offshore","X2"="North Offshore", "X3" = "Inshore")) +
  theme(axis.text.x = element_blank(),panel.grid = element_blank())+
  labs(y="Ancestry (K=3)", x="") 

p1/p2
```

```{r cv, fig.retina=2, fig.width=3.4, fig.height=3.8,out.extra='style="float:left"'}
cv <- read_delim("data/hpc/admixture/CV.txt", col_names = c("CV","value"),delim = " ")
ggplot(cv,aes(x=CV,y=value)) + geom_line() + geom_point(shape=21,fill="white",alpha=1) +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  labs(x="K",y="Cross-validation error")
```




















