Gene Annotations
================

-   [Scan for conserved domains](#scan-for-conserved-domains)
-   [Homology search with BLAST](#homology-search-with-blast)

As a precursor to a range of gene functional analyses, we generated an
annotated gene set for *A. digitifera*. Our annotations are based on
gene models provided by Chuya Shinzato for the version 2 genome. These
annotations were provided as file `Adig-v2.gff` along with a translation
table `20191203c.shinzato-0010_scaff_adig.acclist.txt` to convert
scaffold ids in this file to those in the NCBI assembly,
`GCA_014634065.1_Adig_2.0_genomic.fna`.

First we extracted protein coding sequences and nucleotide sequences for
the longest isoform per gene model. These sequences were used for BLAST
and InterProScan analyses.

``` bash
cgat gff2gff --filter=longest-gene -I annotation_from_chuya/adig-v2.gff -S annotation_longest_gene.gff

gffread -g reference_entryname.fa -y protein.fa annotation_longest_gene.gff
gffread -g reference_entryname.fa -x CDS.fa annotation_longest_gene.gff
```

# Scan for conserved domains

Next we used the
[InterProScan5](https://www.ebi.ac.uk/seqdb/confluence/display/JDSAT/InterProScan+5+Help+and+Documentation#InterProScan5HelpandDocumentation-RESTAPI)
REST API to identify conserved domains in protein translations of all
genes. This analysis was primarily used to provide a set of GO term
assignments based on conserved domains rather than specific genes. We
call these GO terms `ipr_go` to distinguish them from those obtained
from uniprot, which are terms assigned to a specific homologous gene.
The `ipr_go` terms will tend to be less specific but are likely to be
more reliable than those provided by uniprot.

``` bash
python iprscan5.py --multifasta protein.fa \
--maxJobs 29 --useSeqId --email email@address.com \
--outformat tsv
```

# Homology search with BLAST

To identify homologs of *A. digitifera* with high quality functional
annotations we used BLAST\[xp\] to search all genes against the
swissprot uniprot database. After filtering blast results to include
only those with evalue &lt;1e-5, we then selected the best available hit
based on evalue. For all these best hits we then looked up putative gene
names, GO terms, and Kegg information from [Uniprot ID
mapping](https://www.uniprot.org/uploadlists/) from UniprotKB AC/ID to
UniprotKB.

| InterProScan | Uniprot(blast) |
|-------------:|---------------:|
|    0.4554257 |      0.6433592 |

The proportion of genes with GO term annotation
