Placing Western Australian Acropora digitifera in context
================

To understand the relationship between *A. digitifera* from WA and other
populations from this species we combined our data with publicly
available data from (Shinzato et al. 2015). This combined dataset
included all of our samples as well as 16 samples from the Ryukyu
Islands in Japan, randomly chosen from the Shinzato et al dataset. Note
that the Ryukyu Islands is also the region of origin for the A.
digitifera reference genome. Japanese samples downloaded from NCBI and
subjected to the same read quality checks and alignment process as we
applied for our own samples. The resulting bam files were as follows;

| SRA Ref   | Mapping Rate | Coverage Depth |
| --------- | ------------ | -------------- |
| DRR099286 |              |                |
| DRR099287 |              |                |
| DRR099291 |              |                |
| DRR099303 |              |                |
| DRR099319 |              |                |
| DRR099320 |              |                |
| DRR099350 |              |                |
| DRR099351 |              |                |
| DRR099361 |              |                |
| DRR099370 |              |                |
| DRR099377 |              |                |
| DRR099383 |              |                |
| DRR099391 |              |                |
| DRR099398 |              |                |
| DRR099421 |              |                |
| DRR099428 |              |                |

### SNP selection and genotype likelihood calculation

As a precursor to running PCAngsd (see below) we first called SNPs,
filtered sites for quality and calculated genotype likelihoods using
angsd (v 0.933) with the
command.

``` bash
angsd -b all_bam.txt -out all -GL 2 -nThreads 8 -doGlf 2 -SNP_pval 1e-6 -doMajorMinor 1 -doMaf 2 -doCounts 1 -minMaf 0.05 -minInd 82 -minMapQ 20 -minQ 20 -setMinDepth 910 -setMaxDepth 3000 -setMinDepthInd 3
```

The options used in this command were chosen to mimic GATK genotype
calling and SNP filtering in our main analysis as closely as possible.
Specifically, `-GL 2` uses a GATK-like genotype likelihood calculation,
`-minInd 82` selects only sites with data for at least 90% of
individuals.

### PCAngsd

We then used PCAngsd version 1.0 obtained from [github commit
cd7feb5a8a76946acde8beaceb919e59dc8d0298](https://github.com/Rosemeis/pcangsd.git)
to perform a PCA analysis. Note that PCAngsd is designed to handle
datasets with differing sequencing depths as is the case
here.

``` bash
pcangsd.py -beagle all.beagle.gz -selection -snp_weights -pcadapt -admix -tree -maf_save -pi_save -dosage_save -sites_save -threads 32 -out jpwa
```

Using the covariance matrix resulting from the analysis we performed PCA
and visualise the distribution of points according to location as shown
below;

![](10.outgroup_analysis_files/figure-gfm/unnamed-chunk-2-1.png)<!-- -->

**Figure 1:** PCA showing variation between and within samples along
PC1, PC2 and PC3.

![](10.outgroup_analysis_files/figure-gfm/unnamed-chunk-3-1.png)<!-- -->

Tree build with UCE and Exon set (Cowman et al)

![](10.outgroup_analysis_files/figure-gfm/unnamed-chunk-4-1.png)<!-- -->

<div id="refs" class="references">

<div id="ref-Shinzato2015-rz">

Shinzato, Chuya, Sutada Mungpakdee, Nana Arakaki, and Noriyuki Satoh.
2015. “Genome-Wide SNP Analysis Explains Coral Diversity and Recovery in
the Ryukyu Archipelago.” *Sci. Rep.* 5 (December): 18211.

</div>

</div>
