---
title: "Scaffolding to Pseudo-chromosomes with RagTag"
output: 
  github_document
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
library(tidyverse)
library(ggpubr)
```

In order to visualise the genomic position of sites under selection at a Chromosome level we used [RagTag](https://github.com/malonge/RagTag) v1.1.1 [@Alonge2019-yy] to align the *Acropora digitifera* genome to the *Acropora millepora* chromosome-level assembly available from the [Przeworski Lab](https://przeworskilab.com/data/) website.

This was done using the `scaffold` command with default settings;

```bash
ragtag.py scaffold Amil.v2.01.chrs.fasta GCA_014634065.1_Adig_2.0_genomic.fna -t 32
```

This process was able to place 735 of the 955 sequences comprising 405 `r 404511305/415828151` percent of bases. 

For our purposes the most important output from RagTag is the `agp` file which we can use to translate coordinates from the old unplaced assembly to the new scaffolded assembly. Although RagTag provides a python script for translating gff files we needed a more general purpose script to translate simple coordinate lists.  To support this we provide the script `data/hpc/ragtag/translate_coords.py` which is essentially a modified version of the RagTag gff conversion script. 

Examples of the use of this script can be found in [12.pcangsd_selection](12.pcangsd_selection.md) and [13.popgen_stats_angsd](13.popgen_stats_angsd.md)



