---
title: "TMRCA"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(rehh)
```

Since hard selective sweeps lead to dominance of a single haplotype at the swept locus it is possible to estimate the date at which a sweep occurred.  This is because new haplotypes will be introduced via mutation, and recombination will act to break down the length of the region of extended haplotype homozygosity. 

The software [startmrca](https://github.com/jhavsmith/startmrca) is designed to estimate the timing of sweeps based on patterns of haplotype diversity and length that arise as a result of mutation and recombination after a sweep has occurred. 

`startmrca` relies on the concept of a "focal allele" which in-theory should represent the allele that is under selection, however, in-practise it need only represent an allele on the same haplotype as the selected allele in strong linkage disequilibrium with it. 

In order to estimate the timing of selection at a locus we need the following information;
  - The position of the focal allele
  - The identity of the focal allele at this position (0 or 1 for a biallelic SNP)
  - A list of individuals with haplotypes containing the focal allele
  - A list of individuals without the focal allele


## Identifying the focal Allele

To identify the focal allele we look for the position with the maximum z-score from |iHS| or XP-EHH within the interval. 

This was done using [...methods...](jia_how_was_this_done?)

```{r}
inshore_sweeps <- readxl::read_excel("data/hpc/startmrca/Supplementary Tables.xlsx",sheet="Table S9", skip = 1) %>% 
  unite("sweep_id",`Scaffold`,Start,sep = "_",remove = FALSE) 
```


```{r}
#BLFC01000007.1:93469	93469	0.133333	0.0162989	0.0169216	-0.0162829	-1.50666	0
ihs_colnames <- c("locus","position","1_freq","ihh_1","ihh_0","ihs","norm","top1pc")
#xpehh_colnames <- c("chr","position","norm")

ihs <- read_tsv("data/hpc/selection/ihs/inshore/all.ihs.out.norm",col_names = ihs_colnames) %>% 
  select(-position) %>% 
  separate(locus,into = c("chr","position"),sep = ":") %>% 
  mutate(position=as.integer(position))
#xpehh_no <- read_tsv("data/hpc/selection/xpehh/inshore_vs_northoffshore.xpehh.norm",col_names = xpehh_colnames)
#xpehh_so <- read_tsv("data/hpc/selection/xpehh/inshore_vs_southoffshore.xpehh.norm",col_names = xpehh_colnames)
```
Since XP-EHH is a bit complicated we start with the iHS sites

```{r}
inshore_sweeps_ihs <- inshore_sweeps %>% 
  filter(`Sig in`=="iHS") %>% 
  select(chr=`Scaffold`,start=Start, end=End)

inshore_sweeps_ihs_top <- inshore_sweeps_ihs %>% 
#  head(1)  %>% 
  pmap_dfr(function(chr,start,end){
    cat(chr,start,end,"\n")
    chrom <- chr
   sites <- ihs %>%
     filter(chr==chrom) %>%
     filter(between(position,start,end)) %>% 
     slice_max(abs(ihs),n = 1, with_ties = FALSE)
   
    sites %>% 
      mutate(derived_allele = as.integer(ihs>0)) %>% 
      mutate(region = paste(chr,":",start,"-",end,sep="")) %>% 
      mutate(sweep_id = paste(chr,"_",start,sep="")) %>% 
      add_column(start = start-1e5, end = end+1e5) %>% 
      mutate(start = ifelse(start<0,0,start)) %>% 
      select(region,position,derived_allele, sweep_id,chr,start,end) 
  })

write_tsv(inshore_sweeps_ihs_top,"data/hpc/startmrca/inshore_ihs_sweeps.tsv")
```


Now that we have written input files to disk, run the MCMC on HPC using 

```bash
bash 03_run_tmrca.sh
```

# Results

```{r}
estimate_tmrca <- function(result){
  result %>% 
  mutate(n=row_number()) %>% 
  mutate(kept = as.factor(V4)) %>% 
  slice_tail(n=10000) %>% 
  slice_sample(n=100) %>% 
    summarise(tmrca_mean = mean(V1), var = var(V1))
}

read_chain <- function(path){
  tmp <- read_rds(path) 
  tmp$t.chain %>% 
    as.data.frame() %>% 
    mutate(p = path) %>% 
    extract(p,"locus",regex="(B.*).rds")
}

tmrca_data <- list.files("data/hpc/startmrca/results/","*.rds", full.names = TRUE) %>% 
  map_dfr(read_chain)

tmrca_estimates <- tmrca_data %>% 
  group_by(locus) %>% 
  slice_tail(n=10000) #%>% 
  slice_sample(n=100) %>% 
  summarise(tmrca_mean = mean(V1), var = var(V1))

```

```{r}
tmrca_data %>% 
  mutate(t_years = V1*5) %>% 
  group_by(locus) %>% 
  mutate(n=row_number()) %>% 
  slice_tail(n=10000) %>% 
  ungroup() %>% 
  ggplot(aes(y=t_years)) + geom_boxplot(aes(fill=locus))
```


# Experimentation


```{r}
hh <- data2haplohh(hap_file = "data/hpc/startmrca/locus_1_info.vcf",
                   polarize_vcf = FALSE,
                   vcf_reader = "data.table")

furcation <- calc_furcation(hh,mrk = "127808")

pdf("test.pdf",height = 16,width = 10)
plot(furcation,hap.names = hap.names(hh),cex=0.5)
dev.off()
```


```{r}
plot(hh)
```

Infer parameters for the startmrca calculation

```{r}
library(tidyverse)
derived_haps <- which(haplo(hh)[,which(positions(hh)==127808)]==1) %>% names()
anc_haps <- which(haplo(hh)[,which(positions(hh)==127808)]==0) %>% names()

derived_indv <- derived_haps %>% str_remove("_[0-9]$")
anc_indv <- anc_haps %>% str_remove("_[0-9]$")

derived_only_indv <- setdiff(derived_indv,anc_indv) %>% as.data.frame()
anc_only_indv <- setdiff(anc_indv,derived_indv) %>% as.data.frame()

write_tsv(derived_only_indv,"data/hpc/startmrca/derived_indv.txt",col_names = FALSE)
write_tsv(anc_only_indv,"data/hpc/startmrca/anc_indv.txt",col_names = FALSE)
```


We estimate the age of beneficial alleles using [startmrca](https://academic.oup.com/mbe/article/35/4/1003/4816405) 



```{r}
# vcf.file	
# a vcf file that includes genotypes of individuals at genomic region of interest surrounding the putatively selected. A 1 Mb region is recommended (500kb flanking both sides of the seleced allele). This file should contain individuals with and without the selected allele.

locus1 <- run.startmrca(vcf.file="data/hpc/startmrca/locus_1.vcf", 
              rec.file=NULL, 
              mut.rate = 1.2e-8,rec.rate= 1.2e-8, 
              nsel = nrow(derived_only_indv), nanc = nrow(anc_only_indv), 
              chain.length = 50000, 
              proposal.sd = 20, 
              nanc.post = 100, 
              sample.ids = "data/hpc/startmrca/derived_indv.txt", 
              refsample.ids = "data/hpc/startmrca/anc_indv.txt",
              pos = 127808, 
    sel.allele = 1, bed.file = NULL, upper.t.limit = 10000, 
    output.file = NULL)

hist(locus1$t.chain[,1])
```


```{r}
r1 <- read_rds("data/hpc/startmrca/results/BLFC01000074.1_100001.rds")
r1$t.chain %>% as.data.frame() %>% 
  mutate(n=row_number()) %>% 
  mutate(kept = as.factor(V4)) %>% 
  slice_sample(n=1000) %>% 
  ggplot(aes(x=n,y=V1)) + 
  geom_point(aes(color=kept)) 
```


```{r}
r2 <- read_rds("data/hpc/startmrca/results/BLFC01000326.1_1550001.rds")
r2$t.chain %>% as.data.frame() %>% 
  mutate(n=row_number()) %>% 
  mutate(kept = as.factor(V4)) %>% 
  slice_sample(n=1000) %>% 
  ggplot(aes(x=n,y=V1)) + 
  geom_point(aes(color=kept)) 
```

```{r}
r3 <- read_rds("data/hpc/startmrca/results/BLFC01000326.1_1750001.rds")
r3$t.chain %>% as.data.frame() %>% 
  mutate(n=row_number()) %>% 
  mutate(kept = as.factor(V4)) %>% 
  slice_sample(n=1000) %>% 
  ggplot(aes(x=n,y=V1)) + 
  geom_point(aes(color=kept)) 
```


```{r}
r4 <- read_rds("data/hpc/startmrca/results/BLFC01000690.1_2750001.rds")
r4$t.chain %>% as.data.frame() %>% 
  mutate(n=row_number()) %>% 
  mutate(kept = as.factor(V4)) %>% 
  slice_sample(n=1000) %>% 
  ggplot(aes(x=n,y=V1)) + 
  geom_point(aes(color=kept)) 
```


```{r}
r5 <- read_rds("data/hpc/startmrca/results/BLFC01000706.1_350001.rds")
r5$t.chain %>% as.data.frame() %>% 
  mutate(n=row_number()) %>% 
  mutate(kept = as.factor(V4)) %>% 
  slice_sample(n=1000) %>% 
  ggplot(aes(x=n,y=V1)) + 
  geom_point(aes(color=kept)) 
```


```{r}
r6 <- read_rds("data/hpc/startmrca/results/BLFC01000706.1_900001.rds")
r6$t.chain %>% as.data.frame() %>% 
  mutate(n=row_number()) %>% 
  mutate(kept = as.factor(V4)) %>% 
  slice_sample(n=1000) %>% 
  ggplot(aes(x=n,y=V1)) + 
  geom_point(aes(color=kept)) 
```
