---
title: "Demographic model fitting with fastsimcoal"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.retina = 2)
require(tidyverse)
require(ggpubr)
require(cowplot)
library(plyr)
```

We used `fastsimcoal` version 2.7 (the lastest version 2705 which fixed a bug regarding migration estimate in models with growth rate) to estimate demographic parameters based on the site frequency spectrum (SFS). 

## SNP filtering and SFS calculation
We used fastsimcoal 2.7 to infer past demographics by maximising a likelihood function based on the fit of a coalescent model to the genome-wide site frequency spectrum (SFS). We used all samples, except BR_5_121_S125_L004 (see [18.radseq_check](18.radseq_check.md) ) which was likely mislabelled from inshore (29), all from north offshore (20), and all from south offshore (25). To minimise potential bias from linkage and selection, we only used non-genic SNPs, and filtered SNPs with an r2 > 0.3 which left us with 257,314 sites. To utilise the mutation rate in branch length computation, we estimated the monomorphic sites based on the proportional number of mappable sites defined by the SNPable pipeline we used in MSMC analysis. With this set of SNPs, we  then used [easySFS](https://github.com/isaacovercast/easySFS) to generate the folded multidimensional allele frequency spectrum (jAFS) for three populations. 

**Prepare dataset**
```bash
bedtools subtract -a Adigi.filtered.vcf.gz -b CDS.bed | bgzip >Adigi.nongenic.vcf.gz
bcftools +prune -m 0.3 -e'F_MISSING>=0.01' -w 1000 Adigi.nongenic.vcf.gz -o Adigi.nongenic.r03nomiss.vcf.gz
## calculate the number of monomorphic sites
## ([the number of mappability sites (from MSMC)] - [genic sites] )* (257,314/8,714,015) - SNPs
#(250,379,843 - 37,088,136)*(257,314/8,714,015) - 257314
#~ 6040924
## add number to sfs later
```

**Calculate SFS**
```bash
python easysfs.py -i Adigi.nongenic.r03nomiss.vcf.gz -p popfile.txt -a --GQ 20 --dtype int -o sfs --proj=58,40,52
```
## Model Selection

We used a hierarchical approach to model selection, first testing alternative models for the tree topology (divergence), then alternative migration and growth scenarios based on the best topology.

### Divergence
We first tested alternative models of the divergence topology. Among the four alternative models, the model where inshore split first clearly displays the best likelihood and AIC values.

```{r}
read_ll <- function(filename) {
  read_tsv(filename) %>% separate(col = File,"/",into = c("model","run"))
}
n_parameter <- read_csv("data/hpc/demography/fastsimcoal/n_parameters.txt", col_names = c("model","n"))
divergence_ll <- list.files("data/hpc/demography/fastsimcoal/divergence",pattern = ".*\\.param",full.names = T) %>% map_df(read_ll)

divergence_ll <- divergence_ll %>% left_join(n_parameter) %>% 
  dplyr::mutate(aic=2*n-2*(MaxEstLhood/log10(exp(1))), delta_ll= MaxObsLhood-MaxEstLhood) 
  
#divergence_ll %>% group_by(model) %>% slice_max(MaxEstLhood,n=1) %>% select(model, run, n, MaxEstLhood, aic) %>% knitr::kable()

ggplot(divergence_ll, aes(x=model,y=aic)) + geom_boxplot() +
  labs(x="Topology",y="AIC") + scale_x_discrete(labels=c("split_IN"="(IN,(NO,SO))","split_NO"="(NO,(IN,SO))","split_poly"="(IN,NO,SO)","split_SO"="(SO,(NO,IN))"))
```

**Figure 1:** The AIC distribution of four divergence models. Plot shows the best AIC value from each of 100 independent runs for each model.

### Models with growth rate and migration.

Since our population genetic statistics and SMC++ analyses all point toward a bottleneck and recent population growth we added a growth phase to all populations after splitting. We then generated 6 alternative models based on the relative timing of migration as shown in the diagram below. 

```{r}
knitr::include_graphics("figures/tested_models.png")
```

**Figure 2: Graphic illustration of six alternative models.  Each model is given a long name and a short code in brackets. **
 
**Model comparison**

For each of the six models, we performed 100 independent runs and examined the distribution of parameter values and AIC values from the maximum likelihood estimate of each run. 

In two of our models we allowed the migration phase to start independently of population splits (TMIG) (though constrained to be after TDIV1), however we found that these two models converged toward a simpler model, "Contemporary continuous migration IMc" (coded as `growth_rate_SC` internally). This can be seen in the parameter plot below where we see that the parameter `TMIG` converges to TMIG=TDIV1 for the model "Secondary contact SC" and converges to 0 for the model "Early migration EM".  Both scenarios reduce to the model "IMc" in this situation. 

```{r, fig.height=3.6}
#gr_ll <- list.files("data/hpc/demography/fastsimcoal/final",pattern = ".*fixed\\.param",full.names = T) %>% 
gr_ll <- list.files("data/hpc/demography/fastsimcoal/final2705",pattern = ".*\\.param",full.names = T) %>%
  map_df(read_ll) %>% 
  left_join(n_parameter) %>% 
  dplyr::mutate(aic=2*n-2*(MaxEstLhood/log10(exp(1))), delta_ll= MaxObsLhood-MaxEstLhood,Tdiff=TDIV1-TMIG) %>% 
  mutate(model.lab = case_when(
      model == "0.growth_rate_SI" ~ "SI",
      model == "1.growth_rate_IM" ~ "IM",
      model == "3.growth_rate_SC" ~ "IMc",
      model == "4.growth_rate_SC2" ~ "SC",
      model == "6.growth_rate_AM2" ~ "EM",
      model == "7.growth_rate_AM" ~ "AM"
      )
  )

gr_ll %>% 
  pivot_longer(-c(MaxEstLhood,model,run, model.lab),names_to = "param",values_to = "value") %>% 
  filter(param %in% c("TDIV1","TMIG","Tdiff") & model %in% c("4.growth_rate_SC2","6.growth_rate_AM2")) %>% 
  ggplot(aes(x=param,y=value)) +
  geom_violin()+
  geom_jitter(size=1) + 
  facet_wrap(~model.lab,scales = "free") +
  labs(x="",y="Time in generations") + scale_x_discrete(limits=c("TDIV1","TMIG","Tdiff"),labels=c("Tdiff"="TDIV1-TMIG"))
```

**Figure 3: Estimated population split times and migration time (TMIG) in two models where the migration time (TMIG) represents migration onset (SC) or termination (EM)**


Model selection analysis based on AIC for all six models clearly showed that the strict isolation model and ancient migration models were not supported (highest AIC).  Other models, EM, IM, IMc and SC all had a similar distribution of AIC values however since SC and EM effectively reduce to IMc this left only the IM and IMc models as potential candidates for the best model. We decided to proceed with `IMc` as the best model on the basis that it is simpler (one less parameter) than `IM` and had a better overall weighted AIC value. 


```{r}
gr_ll %>% 
  ggplot(aes(x=model.lab,y=aic)) + 
  geom_violin(trim = FALSE,fill="gray") + 
  stat_summary(fun.data=mean_sdl, mult=1, 
                 geom="pointrange", color="darkred",size=.6,fatten=2) +
  theme_test(base_size = 12) + labs(y="AIC",x="")

```

<!--

#### Eventually, I used non-parametric bootstrapping, thus comment out this section

To obtain confidence intervals for the parameters in `IMc` we used a parametric bootstrapping approach. For this, we simulated 100 datasets by modifying the FREQ to DNA in parameter estimate file (`3.growth_rate_SC_maxL.par`) which maximise the likelihood. simulate 257,314 DNA fragments with size of 24bp.

```bash
fsc27 -i 3.growth_rate_SC_boot.par -n100 -j -m -s0 -x -I --foldedSFS -q -c12 -B12
```
Then we perform 20 independent runs for each of the bootstrap data using the same parameters for real data

```bash
fsc27 -t 3.growth_rate_SC_boot.tpl -e 3.growth_rate_SC_boot.est -n200000 -m -M -L50 -C2 -c16 -B16 -q
```
-->

### Non-parametric bootstrapping
To estimate the confidence interval of all the parameters. We did 100 times of sampling with replacement(using [sample](https://github.com/alexpreynolds/sample)) of SNPs in our filtered SNP sets with sample number of loci and created the joint SFS the same way. Next, we performed 20 independent fastsimcoal runs with the same parameters for each bootstapping data set and obtain the range of parameters from the best estimates from those runs.

```bash
grep -v "#" Adigi.v2.nongenic.r03nomiss.SNPs.vcf > temp
grep "#" Adigi.v2.nongenic.r03nomiss.SNPs.vcf > header.txt
cat header.txt <(sample -k 257314 temp -r) > bootstrap.vcf
```

### The distribution of the values of parameters estimated.

```{r}
bs_df <- read_tsv("data/hpc/demography/fastsimcoal/final2705/bootstrap_params.txt") %>% filter(TDIV1<9000)%>%
  dplyr::mutate(bs_id=row_number()) %>% 
  pivot_longer(-c(bs_id,MaxEstLhood,MaxObsLhood), names_to = "param", values_to = "value")
```

**Divergence time range**

```{r}
## To remove outlier
#outliers <- boxplot(bs_df$MaxEstLhood, plot=FALSE)$out
#bs_df[which(! bs_df$MaxEstLhood %in% outliers),]

bs_df %>% filter(param %in% c("TDIV1","TDIV2"))  %>% 
  ggplot(aes(x=param, y= value*5)) + geom_boxplot() + labs(y="Years ago",x="")

```

**Estimated population sizes**

```{r}
bs_df %>% filter(grepl(param, pattern = "^NA")) %>% 
  ggplot(aes(x=param, y= value)) + geom_boxplot() + labs(x="",y="Ne")
```

**Migration rate**

```{r}
bs_df %>% filter(grepl(param,pattern = "MIG")) %>% 
  ggplot(aes(x=param, y= value)) + geom_boxplot() + labs(x="",y="migration rate")
```

**Tables: The point estimate of six models and the parameter 95% confidence intervals obtained under IMc model which were calcualted according to the percentile method.**

```{r }
#creating table s8
gr_ll %>% group_by(model.lab) %>% slice_min(aic,n=1) %>% select(-c(model,run)) %>% 
  pivot_longer(-c(model.lab),names_to = "parameters", values_to = "value") %>% 
  pivot_wider(names_from = model.lab,values_from = value) %>% select(parameters,SI,IM,IMc,SC,EM,AM) %>% knitr::kable()
#%>% write_tsv("table-s8.txt")

#bs_df %>% group_by(param) %>% dplyr::summarise(min=min(value),max=max(value)) %>% write_tsv("table-s8b1.txt")

bs_df %>% group_by(param) %>% dplyr::summarise(lowerbound=quantile(value,probs=c(0.05)),upperbound=quantile(value,probs=c(0.95))) %>% knitr::kable()
```
