Demography with FSC
================

FastSimCoal can fit a demographic model to SFS data and it can also be
used to simulate data under a demographic model. In our model selection
process with FSC we determined the best model and obtained bootstrap
estimates for its parameters (see
[22a.fastsimcoal_fitting](22a.fastsimcoal_fitting.md)). These were then
used as priors to simulate data under the model. An advantage of this is
that it allows us to calculate many summary statistics other than just
the SFS from the data, and determine whether these match or deviate from
our real data.

**Parameter Estimates**

| param   |            lb |            ub |
|:--------|--------------:|--------------:|
| ANCSIZE |  4.068171e+05 |  4.228167e+05 |
| MIGN    |  3.580000e-05 |  7.760000e-05 |
| MIGO    |  1.123000e-04 |  1.798000e-04 |
| MIGS    |  1.789000e-04 |  1.922000e-04 |
| NAI     |  5.295900e+03 |  8.725350e+03 |
| NAN     |  7.864750e+03 |  1.415070e+04 |
| NAOFF   |  1.951930e+04 |  9.997235e+04 |
| NAS     |  4.667000e+03 |  9.055400e+03 |
| NI      |  2.077248e+05 |  3.590960e+05 |
| NN      |  4.687270e+05 |  9.826125e+05 |
| NS      |  4.900837e+05 |  1.065497e+06 |
| RI      | -1.091800e-03 | -6.540000e-04 |
| RN      | -1.382800e-03 | -9.134000e-04 |
| RS      | -1.549100e-03 | -1.033900e-03 |
| TDIV1   |  3.110450e+03 |  4.457200e+03 |
| TDIV2   |  3.498300e+03 |  5.230600e+03 |

**Simulations**

Using these parameters as priors we then run FSC to generate data as
follows;

``` bash
../fsc27_linux64/fsc2702 -t 3.out.growth_rate_SC.tpl -n 1 -e 3.out.growth_rate_SC.est  -E 10 -G x -c 0
```

Where the file
[3.out.growth_rate_SC.tpl](data/hpc/fastsimcoal/growth_rate_SC/3.out.growth_rate_SC.tpl)
and specifies that the simulation should generate 20 independent
chromosomes of length 2mb using a recombination rate of 3.2e-8 and
mutation rate of 1.2e-8. The outputs can be converted to vcf using the
awk script [gen2vcf](data/hpc/fastsimcoal/gen2vcf.awk).

## Calculations based on simulated data

Long runs of homozygosity were calculated using ibdseq in the same way
as for real data .. ie as in [06.ibd_hbd](06.ibd_hbd.md)

PLINK was used to calculate heterozygosity statistics just as in
[04.popgen_stats](04.popgen_stats.md)

And `vk tajima` was used to calculate Tajimaâ€™s D in sliding windows as
in [04.popgen_stats](04.popgen_stats.md)

Admixture under the simulated model was assessed using `ADMIXTURE` in
the same manner as for real data (as in
[05.population_structure](05.population_structure.md)). Since the
simulated data only included 20 chromosomes of size 2Mb the number of
SNPs used was much lower than in the real data after filtering, however,
all individuals still had assignment values of almost 100% to the
relevant cluster. Results are shown visualised for a single simulation
run in the figure

![](22b.fastsimcoal_sim_files/figure-gfm/unnamed-chunk-6-1.png)<!-- -->
